service/api_logic/user_logic.py

```diff
@@ -128,4 +135,64 @@ async def create_access_token_response(self, user):
         response_json = jsonify(response)
         set_access_cookies(response_json, access_token)
 
-        return response_json
\ No newline at end of file
+        return response_json
+
+
+    def add_preferences(self, dto: UpdateUserPreferencesDTO):
+        if not dto.user_id or not isinstance(dto.preferences, list):
+            raise NotUserIdOrPreferencesError()
+
+        new_dto_by_type_of_preference = self.dto_for_type_of_preference(dto)
+
+        existing_sports = self._preferences_dal.get_all_sport_preference_indexes()
+        if dto.type == SPORT_TYPE:
+            valid_preferences = set([sport_id for sport_id in dto.preferences if sport_id in existing_sports])
+
+        if dto.type == TEAM_TYPE:
+            valid_preferences = set([sport_id for sport_id in dto.preferences])
+
+        if not valid_preferences:
+            raise IncorrectPreferencesError()
+
+        self._preferences_dal.delete_all_user_preferences(new_dto_by_type_of_preference, dto)
+        self._preferences_dal.add_user_preferences(new_dto_by_type_of_preference, dto, valid_preferences)
```

Comment:
What if new preferences differ just by one preference from a previous set? Will we still remove everything and add them from scratch?